### PR TITLE
VimL (Vimscript) to Lua translator

### PR DESCRIPTION
**Pull request by ZyX-I**
This is the current work on VimL to lua translator.

TODO:
- [x] expression parser
- [x] expression parser tests
- [x] expression parser documentation
- [ ] Ex commands parser
  - [x] Structure commands (branching, cycles, function definitions, try block)
  - [x] Modifier command family
  - [x] `:*do` command family (`argdo`, `bufdo`, etc)
  - [x] `:edit` command family
  - [x] Commands that use patterns (`:s`, `:g`, …)
  - [x] Other commands
  - [ ] Save information about “tokens” present on the command-line, to be used for highlighting and more deep analysis then current AST allows later
  - [ ] As it processes parsing, save information about completion context to use, to be used in place of `set_one_cmd_context`
- [x] Ex commands parser tests
  - [x] Structure commands (branching, cycles, function definitions, try block)
  - [x] Modifier command family
  - [x] `:*do` command family (`argdo`, `bufdo`, etc)
  - [x] `:edit` command family
  - [x] Commands that use patterns (`:s`, `:g`, …)
  - [x] Other commands
- [ ] Ex commands parser documentation
  - [ ] (partial) Present functions documentation
  - [ ] Developer documentation
- [ ] VimL to lua translator
  - [x] Complex variable names (curly braces names)
  - [x] Block commands
    - [x] `:if` block
    - [x] `:while` block
    - [x] `:for` block
    - [x] `:try` block
    - [x] flow control commands: `break` and `continue`
  - [x] Assignment support
    - [x] `:let` assignment
    - [x] `:let` slice assignment
    - [x] `:function` definitions
    - [x] `:let` modifying assignment (`+=`, `-=`, `.=`)
    - [x] `:unlet`
    - [x] `:delfunction`
  - [ ] `:command` definitions
  - [x] Error handling (saving error positions)
  - [ ] Commands taking other commands as arguments (modifier, `:*do`, `:e +cmd`, …)
  - [ ] Other commands
- [ ] VimL to lua translator tests
  - [x] Complex variable names (curly braces names)
  - [ ] Block commands
    - [x] `:if` block
    - [x] `:while` block
    - [ ] (partial: no tests for modifying lists during iteration) `:for` block
    - [x] `:try` block
    - [x] flow control commands: `break` and `continue`
  - [ ] Assignment support
    - [ ] (partial: only local and global scopes) `:let` assignment
    - [x] `:let` slice assignment
    - [ ] (partial) `:function` definitions
    - [ ] (partial: local and global scopes) `:let` modifying assignment (`+=`, `-=`, `.=`)
    - [ ] (partial) `:unlet`
    - [ ] (indirect) `:delfunction`
  - [ ] `:command` definitions
  - [ ] Commands taking other commands as arguments (modifier, `:*do`, `:e +cmd`, …)
  - [ ] Other commands
- [ ] VimL to lua translator documentation
  - [x] Present functions documentation
  - [ ] Developer documentation
- [ ] Backend lua module
  - [x] Core types
    - [x] Scalar types
    - [x] Container types
      - [x] Locks
      - [x] Type literals
    - [x] Function references
  - [x] Operator support
    - [x] Arithmetic operators (`+-*/%`)
    - [x] String concatenation
    - [ ] Pattern matching
    - [x] Comparison operators (less/greater then (or equal to), equality)
    - [x] Identity operators (is/isnot)
    - [x] Type coercions when using operators
  - [x] Subscripting support
    - [x] `[key]` dictionary, list, string and number subscripting
    - [x] `[idx1:idx2]` list, string and number slicing
    - [x] `(args)` function call
  - [ ] User command calls
  - [x] Error handling
  - [ ] Built-in functions implementation
    - [x] Implemented functions: `string()`, `call()`, `type()`, `copy()`, `deepcopy()`
    - [ ] Partially implemented functions: `function()`
    - [ ] Other functions were not implemented
  - [ ] User functions support
    - [x] `:function` definitions
    - [x] Dictionary functions
    - [x] Defining functions right inside a dictionaries
    - [x] Varargs functions
    - [ ] Range functions
    - [x] Functions with `abort` modifier
  - [ ] Built-in commands implementation
  - [ ] User commands support
  - [ ] Translations support
- [ ] Backend lua module tests
  - [x] Core types
    - [x] Scalar types
    - [x] Container types
      - [x] Locks
      - [x] Type literals
    - [x] Function references
  - [ ] Profiling
  - [ ] Operator support
    - [x] Arithmetic operators (`+-*/%`)
    - [x] String concatenation
    - [ ] Pattern matching
    - [ ] (partial: no check whether &ic option applies) Comparison operators (less/greater then (or equal to), equality)
    - [ ] (partial: no check whether &ic option applies) Identity operators (is/isnot)
    - [x] Type coercions when using operators
  - [ ] Subscripting support
    - [x] `[key]` dictionary, list, string and number subscripting
    - [x] `[idx1:idx2]` list, string and number slicing
    - [ ] `(args)` function call
  - [ ] User command calls
  - [ ] Error handling
  - [ ] Built-in functions implementation
    - [x] Tested functions: `string()`, `copy()`, `deepcopy()`
    - [ ] Partially or indirectly tested functions: `function()`
    - [ ] Other functions were not tested
  - [ ] User functions support
    - [x] `:function` definitions
    - [x] Dictionary functions
    - [x] Defining functions right inside a dictionaries
    - [x] Varargs functions
    - [ ] Range functions
    - [x] Functions with `abort` modifier
  - [ ] Built-in commands implementation
  - [ ] User commands support
- [ ] Backend lua module documentation
- [ ] Old implementation replacement (need to replace some entry point functions)
- [ ] Old implementation removal (need to remove old backend functions that were replaced by lua code)
- [ ] vim/neovim incompatibilities documentation (ref #387)

### CODE REVIEW

**Comment by ZyX-I** [comment]
Finally added the last (:syntax) command parsing support. The next plan is:

1. Do a bit of refactoring (I have actually happened came up with a better architecture for the parser, but doing such kind of things *now* is not a good idea, so only a tiny bit).
2. Implement some TODOs/FIXMEs, fix some known bugs.
3. Fix travis tests.
4. Proceed with #1536.

**Comment by Jookia** [comment]
Amazing work!

**Comment by ZyX-I** [comment]
Finally at least clang and clint tests were fixed.

**Comment by ZyX-I** [comment]
I love gcc `-Wconversion`:

```C
/home/travis/build/neovim/neovim/src/nvim/viml/testhelpers/parser.c:186:24: error: conversion to ‘char’ from ‘int’ may alter its value [-Werror=conversion]

182   do {
183     uint8_t digit = (uint8_t) ((offset >> ((i - 1) * 4)) & 0xF);
184     const char s[] = {(digit < 0xA
185                        ? (char) ('0' + digit)
186                        : (char) ('A' + (char) (digit - 0xA)))};
187     if (write(s, 1, 1, cookie) != 1) {
188       return FAIL;
189     }
190   } while (--i);
```

Where did this thing find fucking int? And I don’t have this warning with gcc-4.8.3.

**Comment by ZyX-I** [comment]
And here as well:

```C
/home/travis/build/neovim/neovim/src/nvim/viml/testhelpers/parser.c:186:24: error: conversion to ‘char’ from ‘int’ may alter its value [-Werror=conversion]

182   do {
183     uint8_t digit = (uint8_t) ((offset >> ((i - 1) * 4)) & 0xF);
184     const char s[] = {(digit < 0xA
185                        ? (char) ('0' + digit)
186                        : (char) ((char) 'A' + (char) (digit - (uint8_t) 0xA)))};
187     if (write(s, 1, 1, cookie) != 1) {
188       return FAIL;
189     }
190   } while (--i);
```

. This is insane.

**Comment by Oberon00** [comment]
> Where did this thing find fucking int?

The literals (e.g. `1`) have the type `int`. Also (some?) operators promote their operands before performing the computation.

**Comment by tarruda** [comment]
I have finally took some time for a quick read through the code and have to say: Amazing work @ZyX-I, this will be by far the biggest code improvement to Neovim.

I'm still finding my way around the code, but I've seen that you are implementing builtin functions and commands in src/nvim/viml/executor/vim.lua. That seems to be the kind of horizontal work that could be split across multiple contributors. Correct me if I'm wrong, but porting a single builtin function or command consists of the following tasks:

- Ensure the existence of any API functions(src/nvim/api) required to implement the function or command(eg:  the `:append` command can probably be implemented on top of `buffer_set_line_slice`)
- Implement the function or command in src/nvim/viml/executor.vim.lua(perhaps split this file into functions.lua, commands.lua, etc). If an API function is required, it will be available in the "api" object(wrappers generated by msgpack-gen.lua and added by `nlua_add_api_functions`, though I haven't seen you use this object in vim.lua)
- Add some tests

This is something anyone can do if provided with instructions/examples of where/how to add the functions, tests, etc.

If you push this branch to Neovim repo and send a PR that implements a single command or function, we could use it as an example for distributng entry-level tasks.

**Comment by ianks** [comment]
@tarruda I think that's a great idea. I've been eager to get my hands dirty and would like to start with something like this.

**Comment by andrewrk** [comment]
This looks like a lot of work that is sitting here bit rotting.

**Comment by tarruda** [comment]
> This looks like a lot of work that is sitting here bit rotting.

This is an unmeasurable amount of work, but @ZyX-I already did a lot: A complete parser that is capable of recognizing a good portion of the current vimL. Note that there's not specification for vimL, and much of its current behavior is actually determined by bugs and consequences of it is implemented(no parser just a "string executer", so just a well-designed parser is already a huge win.

I haven't put any focus on this yet because I know there are more urgent refactorings that need to get out of the way before we can hit vimL that hard, but I'm sure this will eventually continue and we'll pick up where @ZyX-I left off.

**Comment by aisk** [comment]
Many thanks to @ZyX-I

**Comment by tarruda** [comment]
> Can we just merge the parser to master, although it is not used now, but we can use this as a tool to write plugins or other stuffs?

The parser and translator by themselves are not very useful for plugins. Right now I think the parser and translator are mostly complete, but a lot of work still remains: Implementing the runtime which connects the generated lua code to the editor API. This includes implementing the majority of ex commands and vimL functions on top of the lua bindings to the API.

**Comment by ZyX-I** [comment]
There are two major problems with this PR:

1. I am pretty sure I have chosen wrong architecture for the parser. Specifically current parsing results do not allow me to highlight the command-line, requiring huge refactoring to do this. Some (smaller) refactoring will be required to use them for completion.
2. I have actually no idea how to make luajit code fast enough.

I know how to fix 1, but have not enough time and mental force to do something as huge again.

(Also it looks like it is better to squash the whole history again: with such big changes to Neovim there is not much value in keeping commits which are not going to work after rebase.)

**Comment by ZyX-I** [comment]
(Old history is in branch `luaviml-old-6`.)

**Comment by ZyX-I** [comment]
Though no, there is a way to fix 1 without huge refactoring.

**Comment by tarruda** [comment]
> I am pretty sure I have chosen wrong architecture for the parser. Specifically current parsing results do not allow me to highlight the command-line, requiring huge refactoring to do this. Some (smaller) refactoring will be required to use them for completion.

Even so, the fact that you wrote a proper parser will possibly let us extract a grammar that can be fed to another parser generator in a second iteration. Parser generators are very sophisticated and give the user great flexibility in what information is exposed by the parser.(Lua's own PEG library, lpeg is worth looking into)

> I have actually no idea how to make luajit code fast enough.

It is possible to make some subsets of vimL as fast as pure luajit code through analysis of the syntax tree created by the parser. For example, consider the following vimL function:

```vim
function Sum(list)
  let rv = 0
  for i in a:list
    let rv += i
  endfor
  return rv  
endfunction
```

If `list` is guaranteed to be a list, the above could be translated without error into the following lua function:

```lua
function Sum(list)
  local rv = 0
  for i = 1, #items do
    rv = rv + items[i]
  end
  return rv
end
```

Since vimL is dynamic, we cannot make such guarantees at parse time, so I imagine the actual generated function is a lot more complicated and requires calling in the runtime to create scopes, variables, error checking and so on. But if before executing the function you examine the parameter and determine it is a list, then the above version of the function can be used.

This is in essence, the basic idea of how JIT compilers specialize and generate fast native code for dynamic languages like lua or javascript: They generate code that assume certain types for parameters, and if the parameters match the expectations, the optimized code is executed. If not, it falls back to the slower translated function that uses runtime API.

VimL would probably require going a few steps further since executing strings at runtime is much more common than in other dynamic languages. Consider for example, this modification to the `Sum` function that illustrates the problem:

```vimL
function Sum(list)
  if type(a:list) != type([])
    exec 'let g:'.g:_wrong_sum_calls_var_name.' += 1'
  else  
    let rv = 0
    for i in a:list
      let rv += i
    endfor
    return rv  
  endif
endfunction
```

It enhances the function to report to a global variable(user-determined) how many times `Sum` was called with wrong(non-list) parameter. At first sight it may seem it is not possible to optimize this function because of the dynamic code execution, but if before translation the function is factored like this:

```vimL
function! SumTaintedBranch1()
  exec 'let g:'.g:_wrong_sum_calls_var_name.' += 1'
endfunction

function! Sum(list)
  if type(a:list) != type([])
    call SumTaintedBranch1()
  else
    let rv = 0
    for i in a:list
      let rv += i
    endfor
    return rv  
  endif
endfunction
```

Then most of the code could still be optimized. This kind of factoring could be performed with simple manipulations of the AST: If a block of code is "tainted" we extract it into another artificial function that is examined and optimized separately.

**Comment by torpak** [comment]
Regarding 2: making it work is the hard part. Making it fast can be done in a second step.

**Comment by ZyX-I** [comment]
About 1: I think you got the idea from the last commit. Completion may be based on last saved highlight token type.

**Comment by andrewrk** [comment]
@tarruda I thought the point of this was to factor VimL out of the core; instead compiling VimL into Lua code that used the msgpack API. Then you could run it in actual LuaJIT. So the problem of optimizing generated lua code should not be part of neovim but left up to the LuaJIT project.

**Comment by ZyX-I** [comment]
@andrewrk LuaJIT has not enough data to perform some optimizations.

**Comment by thomkeh** [comment]
tarruda wrote:

> Even so, the fact that you wrote a proper parser will possibly let us extract a grammar that can be fed to another parser generator in a second iteration.

So maybe a repository could be created where people can contribute to this goal while the main neovim repository focuses on more pressing issues?

**Comment by tarruda** [comment]
> @tarruda I thought the point of this was to factor VimL out of the core; instead compiling VimL into Lua code that used the msgpack API. Then you could run it in actual LuaJIT. So the problem of optimizing generated lua code should not be part of neovim but left up to the LuaJIT project.

Lua and Viml have different semantics, so the vimL->lua translator needs a runtime component to work properly. LuaJIT is only good at optimizing "normal" lua code, so to access its true potential we must have our own optimizer that is capable of identifying which subsets of vimL that can be safely translated into normal lua code without touching the vimL runtime.

But as @torpak commented before, we don't need this optimizer now, There's already a direct benefit to implementing viml on top of lua: Greatly improved code maintenance.

**Comment by stefanpenner** [comment]
:+1: Keep up the awesome work here, fun to see this evolve :)

**Comment by simone-trubian** [comment]
Hi guys I was thinking of helping out with this PR, is the list made by @ZyX-I at the top of this thread still valid? My knowledge of this project is limited but I could start with some minor task like docs or tests.

**Comment by ZyX-I** [comment]
@simone-trubian AFAIR it was more or less valid at the time I took a long pause (which resulted in squashing everything again), and last commits are more about refactoring and should not have changed the list. Though I need to add a few more entries there.

**Comment by simone-trubian** [comment]
@ZyX-I that's good news, any easy / boring part you'd like me to get started from?

**Comment by ZyX-I** [comment]
@simone-trubian I think that the most needed is the new addition (the one about saving completion contexts). There are no examples yet, but once it is setup it will be definitely

1. Easy.
2. Boring (needs to modify each and every command).
3. Necessary (I do not like the idea of keeping `set_one_cmd_context` which has many code common with `do_one_cmd` because it *also* does parsing; we may cope better by doing parsing *once*).

. Needs an additional `struct` inner to `CommandParserResult` struct.

There are other easy and boring jobs (mostly rewriting every command in lua), but I still am not sure that this will look exactly this way: specifically with the rewriting I am not sure that I will not decide to eventually switch to keeping C functions, writing stack-based VM specifically for VimL or something like this. Another example is adding tokens (see https://github.com/ZyX-I/neovim/commit/3f8e323277b71abb3eef013eecdd8f24829391c6, though after refactoring tokens were planned to be attached to `CommandParserResult` struct as well): I am not sure when and whether somebody will write something like fish highlighting for Vim command-line which is the main reasoning behind this change (“fish” is “Friendly Interactive SHell”, has some nice interactive features out of the box including highlighting as you type, though I still prefer zsh for fish being too limited for my taste; another example of highlighted shell is bpython: if you ever worked with one or the other you will understand why I want this in Vim command line).

---

Though there is one thing from #1536 which will definitely will be needed, and should be rather easy: rewriting each and every `ex_*` function to use only msgpack API functions (note: called directly from C like any other C function). But without previous steps you mostly can only make sure that all necessary functionality is available through the API which is also easy, boring and necessary: making API richer will make not only translator easier to implement, but will also make happy external plugin authours which do not want to touch VimL. I think this has bigger priority then saving translator contexts (especially given that needed refactoring is not yet written). This has also big advantage of being very easily split into a number of small PRs.

**Comment by simone-trubian** [comment]
Hi Nikolai

Sorry for the late reply. I'm still on. Just getting started to wrap my
head around #1536 <https://github.com/neovim/neovim/issues/1536>. Be back
soon with an action plan.

Simone

**Comment by ticki** [comment]
What is the state of this?

**Comment by justinmk** [comment]
@ticki The checklist in the first post on this page shows the exact status.

File: src/nvim/viml/executor/vim.lua [hunk 1]
```diff
@@ -0,0 +1,2296 @@
...
+      ret[get_string(state, key)] = val
+    end
+    return ret
+  end),
+-- {{{4 Modification support
+-- {{{4 Assignment support
+  assign_subscript = function(state, dct, dct_position, key, key_position, val)
+    if key == '' then
+      return err.err(state, key_position, true,
+                     'E713: Cannot use empty key for dictionary')
```
Comment (brcolow) [inline src/nvim/viml/executor/vim.lua]: This was changed in https://github.com/vim/vim/commit/0921ecff1c5a74541bad6c073e8ade32247403d8

**Comment by ticki** [comment]
@justinmk It shows what's complete, not what the state is: Is this PR still active? Is any progress happening or is it temporary halted? etc.

**Comment by justinmk** [comment]
@ticki Parts of this PR have been re-used in other PRs, including https://github.com/neovim/neovim/pull/5119.

VimL-to-Lua translation is on hold indefinitely because of performance concerns. In 0.3 we're going to focus on shipping Lua as a built-in scripting alternative (note: as always, VimL will continue to be supported).

**Pull request closed without merging**